From: Aryan <admin@reticentadmin.com>
Date: 2025-02-15
Subject: [PATCH] Add hardening patches to chromium's ebuild.

This patch implements hardening patches from Trivalent and adds musl support
to the existing chromium ebuild.

Copyright (c) 2025 Aryan
SPDX-License-Identifier: BSD-3-Clause

Version: 1.0.0

--- chromium.ebuid
+++ chromium.ebuid
@@ -31,7 +31,7 @@
 HOMEPAGE="https://www.chromium.org/"
 PPC64_HASH="deefc994ce2d31faf6d27f5e81782e039c663aed"
 PATCH_V="${PV%%\.*}-2"
-SRC_URI="https://chromium-tarballs.distfiles.gentoo.org/${P}-linux.tar.xz
+SRC_URI="https://commondatastorage.googleapis.com/chromium-browser-official/${P}.tar.xz -> ${P}-linux.tar.xz
 		https://gitlab.com/Matt.Jolly/chromium-patches/-/archive/${PATCH_V}/chromium-patches-${PATCH_V}.tar.bz2
 	test? (
 		https://chromium-tarballs.distfiles.gentoo.org/${P}-linux-testdata.tar.xz
@@ -417,6 +417,15 @@
 			die "Failed to remove default visibility nightly option"
 	fi

+	# If the user has musl libc, remove the libatomic dependency.
+	if use elibc_musl; then
+		PATCHES+=( "${FILESDIR}/remove-libatomic.patch" )
+		eapply "${FILESDIR}/musl/"
+	fi
+
+	# Apply hardening patches.
+	eapply "${FILESDIR}/hardening/"
+
 	default

 	# Not included in -lite tarballs, but we should check for it anyway.
@@ -952,10 +961,13 @@
 	# The sysroot is the oldest debian image that chromium supports, we don't need it
 	myconf_gn+=" use_sysroot=false"

-	# Use in-tree libc++ (buildtools/third_party/libc++ and buildtools/third_party/libc++abi)
-	# instead of the system C++ library for C++ standard library support.
-	# default: true, but let's be explicit (forced since 120 ; USE removed 127).
-	myconf_gn+=" use_custom_libcxx=true"
+	# Don't use in-tree libc++ for musl (buildtools/third_party/libc++ and buildtools/third_party/libc++abi),
+	# instead use system C++ library for C++ standard library support.
+        if use elibc_musl; then
+	    myconf_gn+=" use_custom_libcxx=false"
+        else
+            myconf_gn+=" use_custom_libcxx=true"
+        fi

 	# Disable pseudolocales, only used for testing
 	myconf_gn+=" enable_pseudolocales=false"
@@ -1109,6 +1121,15 @@
 		else
 			myconf_gn+=" is_cfi=${use_lto}"
 		fi
+
+		# Disable CFI on musl.
+		# https://github.com/12101111/overlay/issues/41.
+		if use elibc_musl; then
+			myconf_gn+=" is_cfi=false"
+		else
+			myconf_gn+=" is_cfi=${use_lto}"
+		fi
+
 		# Don't add symbols to build
 		myconf_gn+=" symbol_level=0"
 	fi
